elastic/kibana
# [Lens] Implement deep linking and embedding

This PR explores ways to implement deep linking and embedding in Lens.

It includes a small example application which embedds a config inlined in the consuming app with two buttons to randomly change the color of the chart and another one to take the current configuration and make it editable in Lens.

* Every state can easily made editable by the user
* Everything possible with Lens can be accessed this way
* Works with saved objects and inline configs

![Lens embedding](https://user-images.githubusercontent.com/1508364/100359825-605c7900-2ff8-11eb-9a78-6f3ad3a4cc4a.gif)

To test:
* run `yarn start --run-examples`
* Install logs sample data and make sure the logs index pattern is the default one
* Go to `/app/embedded_lens_example`

## Consumer side

This PR adds two things to the lens start contract: 
* `EmbeddableComponent: React.ComponentType<LensProps>` A component usable in your application which takes either a persisted state of a Lens visualization (technically it's the embeddable input, but there's very little difference) or a saved object id of a Lens visualization. The component loads the Lens bundle and uses the Lens embeddable implementation to render the visualization very similar to how it's working in a dashboard. If the input changes, the visualization is rendered again.
* `navigateToPrefilledEditor: (input: LensEmbeddableInput) => void` Takes a Lens embeddable input (again persisted state or saved object id) and navigates to Lens with this state preloaded

`LensProps['attributes']` is properly typed for the individual state shapes, so it's possible to statically check consumers of this API.

`x-pack/examples/embedded_lens_example/public/app.tsx` contains the example plugin using these APIs

## Internals

`x-pack/plugins/lens/public/editor_frame_service/embeddable/embeddable_component.tsx` and `x-pack/plugins/lens/public/editor_frame_service/embeddable/embeddable_component_loader.tsx` provide a very thin wrapper around the embeddable acting as a lazy-loading mini embeddable container, passing through the input via props.

`navigateToPrefilledEditor` simply uses `EmbeddableStateTransfer` to kick off the navigation like the edit action on dashboard does.

Normally the `attributes` of the Lens embeddable input are typed very loose, this PR makes them specific for all existing visualizations:
```
  attributes:
    | LensAttributes<'lnsXY', XYState>
    | LensAttributes<'lnsPie', PieVisualizationState>
    | LensAttributes<'lnsDatatable', DatatableVisualizationState>
    | LensAttributes<'lnsMetric', MetricState>;
```

Once we make it possible to register visualizations from outside of the Lens plugin, we can use an approach similar to UI actions to extend a central interface containing all existing pairs to keep this level of type-safety.

## Problems

* Exposes the internal types - if we treat changes here as breaking change, we would be severly limited in adding/changing stuff in the state. For the beginning pne option would be to explicitly mention this in the docs and mark this API as experimental so consumers are aware the integration can break easily. As it's a single CI build I don't think we will have problems with Kibana-internal integrations this way. Another option would be to introduce a compatibility layer for this which maintains BWC for a major version if that's necessary, but I would wait with that until it actually becomes a problem
* Completely re-initalizes the state on each change - the way the code to turn a document into an expression is set up right now, it does the init routine of the datasource on every change. This can be improved by keeping the initialized state around in the embeddable so we don't have to initialize from scratch on every small change.
* Right now the state transfer moving the state into the Lens app requires an "originatingAppId" - in this PR it's set to empty string which doesn't work correctly everywhere. This can be solved cleanly by making `originatingAppId` optional

## Extensions

* It would be cool to have a bunch of helper functions to construct the Lens state to make it simpler to use
